reload does not work for uvicorn.workers.UvicornWorker

this is my commandline`
gunicorn -k uvicorn.workers.UvicornWorker --reload-engine=inotify   --workers=1 --log-level debug  --access-logfile - --error-logfile - --disable-redirect-access-to-syslog --reload  main:app`

note that i have tested with both master branch of gunicorn as well as last stable release.


This is the code im using
```

from fastapi import FastAPI

app = FastAPI()


@app.get("/")
async def read_root():
    # raise 1
    return {"Hello": "World"}


@app.get("/items/{item_id}")
async def read_item(item_id: int, q: str = None):

    return {"item_id 2": item_id, "q": q}

```

reload does not pickup when i make changes to the app. im using vscode as an editor

I've seen it, but it actually doing nothing.
```gunicorn main:app --reload -k uvicorn.workers.UvicornWorker```

However, if I add `inotify`:
```
gunicorn main:app --reload --reload-engine inotify -k uvicorn.workers.UvicornWorker
[2021-03-17 10:43:49 +0100] [31332] [INFO] Starting gunicorn 20.0.4
[2021-03-17 10:43:49 +0100] [31332] [INFO] Listening at: http://127.0.0.1:8000 (31332)
[2021-03-17 10:43:49 +0100] [31332] [INFO] Using worker: uvicorn.workers.UvicornWorker
[2021-03-17 10:43:49 +0100] [31333] [INFO] Booting worker with pid: 31333
[2021-03-17 10:43:49 +0100] [31333] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/Users/alan/Downloads/fastapi/venv/lib/python3.9/site-packages/gunicorn/arbiter.py", line 583, in spawn_worker
    worker.init_process()
  File "/Users/alan/Downloads/fastapi/venv/lib/python3.9/site-packages/uvicorn/workers.py", line 63, in init_process
    super(UvicornWorker, self).init_process()
  File "/Users/alan/Downloads/fastapi/venv/lib/python3.9/site-packages/gunicorn/workers/base.py", line 132, in init_process
    self.reloader = reloader_cls(extra_files=self.cfg.reload_extra_files,
TypeError: __init__() got an unexpected keyword argument 'extra_files'
[2021-03-17 10:43:49 +0100] [31333] [INFO] Worker exiting (pid: 31333)
[2021-03-17 10:43:49 +0100] [31332] [INFO] Shutting down: Master
[2021-03-17 10:43:49 +0100] [31332] [INFO] Reason: Worker failed to boot.
```
It doesn't work at all. Is there anything special to do about `inotify`? It seems a feature for Linux kernels, I'm running `macOS Big Sur`. Anyway I did `pip install -U uvloop httptools inotify`.

So I tested on Linux (Ubuntu 18.04) as well, failed. My whole setup (you need Python 3.7 or higher):
```bash
mkdir fastapi
cd fastapi
```
create file `main.py`
```python
from typing import Optional
from fastapi import FastAPI
from pydantic import BaseModel

class Item(BaseModel):
    name: str
    description: Optional[str] = None
    price: float
    tax: Optional[float] = None

app = FastAPI()

@app.post("/items/")
async def create_item(item: Item):
    return item
```
Now, set python venv:
``` bash
python3 -m venv venv
source venv/bin/activate
pip install -U fastapi gunicorn pip wheel uvicorn uvloop httptools
```
Standard way work as expected when changing `main.py`:
```
uvicorn main:app --reload
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     Started reloader process [53638] using statreload
INFO:     Started server process [53640]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
WARNING:  StatReload detected file change in 'main.py'. Reloading...
INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [53640]
INFO:     Started server process [53758]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
```
Now, try with basic `gunicorn`, inducing a change that would cause `main.py` to failed (like a typo `returns item`):
``` bash
gunicorn main:app --reload -k uvicorn.workers.UvicornWorker
[2021-03-17 11:43:37 +0000] [95542] [INFO] Starting gunicorn 20.0.4
[2021-03-17 11:43:37 +0000] [95542] [INFO] Listening at: http://127.0.0.1:8000 (95542)
[2021-03-17 11:43:37 +0000] [95542] [INFO] Using worker: uvicorn.workers.UvicornWorker
[2021-03-17 11:43:37 +0000] [95544] [INFO] Booting worker with pid: 95544
[2021-03-17 11:43:37 +0000] [95544] [INFO] Started server process [95544]
[2021-03-17 11:43:37 +0000] [95544] [INFO] Waiting for application startup.
[2021-03-17 11:43:37 +0000] [95544] [INFO] Application startup complete.
[2021-03-17 11:44:02 +0000] [95544] [INFO] Worker reloading: /home/awilter/fastapi/main.py modified
```
So, it may catch a "change", but it's not reloaded.

Using, `inotify`:
``` bash
gunicorn main:app --reload --reload-engine=inotify -k uvicorn.workers.UvicornWorker
[2021-03-17 11:49:58 +0000] [99660] [INFO] Starting gunicorn 20.0.4
[2021-03-17 11:49:58 +0000] [99660] [INFO] Listening at: http://127.0.0.1:8000 (99660)
[2021-03-17 11:49:58 +0000] [99660] [INFO] Using worker: uvicorn.workers.UvicornWorker
[2021-03-17 11:49:58 +0000] [99662] [INFO] Booting worker with pid: 99662
[2021-03-17 11:49:58 +0000] [99662] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/home/awilter/fastapi/venv/lib/python3.8/site-packages/gunicorn/arbiter.py", line
 583, in spawn_worker
    worker.init_process()
  File "/home/awilter/fastapi/venv/lib/python3.8/site-packages/uvicorn/workers.py", line
63, in init_process
    super(UvicornWorker, self).init_process()
  File "/home/awilter/fastapi/venv/lib/python3.8/site-packages/gunicorn/workers/base.py",
 line 132, in init_process
    self.reloader = reloader_cls(extra_files=self.cfg.reload_extra_files,
TypeError: __init__() got an unexpected keyword argument 'extra_files'
[2021-03-17 11:49:58 +0000] [99662] [INFO] Worker exiting (pid: 99662)
[2021-03-17 11:49:58 +0000] [99660] [INFO] Shutting down: Master
[2021-03-17 11:49:58 +0000] [99660] [INFO] Reason: Worker failed to boot.
```
As for `watchgod`, I don't know how to use it, doing simply `--reload-engine=watchgod` does not work.

--- a/THANKS
+++ b/THANKS
@@ -15,6 +15,7 @@ Alex Gaynor <alex.gaynor@gmail.com>
 Alex Robbins <alexander.j.robbins@gmail.com>
 Alexandre Zani <alexandre.zani@gmail.com>
 Alexis Le-Quoc <alq@datadoghq.com>
+Ali Sadeghi Far <booxgitboox@gmail.com>
 Anand Chitipothu <anandology@gmail.com>
 Andreas Stührk <andy-python@hammerhartes.de>
 Andrew Burdo <zeezooz@gmail.com>

--- a/gunicorn/arbiter.py
+++ b/gunicorn/arbiter.py
@@ -581,7 +581,7 @@ class Arbiter(object):
             self.log.info("Booting worker with pid: %s", worker.pid)
             self.cfg.post_fork(self, worker)
             worker.init_process()
-            sys.exit(0)
+            os._exit(0)
         except SystemExit:
             raise
         except AppImportError as e:
@@ -589,12 +589,12 @@ class Arbiter(object):
                            exc_info=True)
             print("%s" % e, file=sys.stderr)
             sys.stderr.flush()
-            sys.exit(self.APP_LOAD_ERROR)
+            os._exit(self.APP_LOAD_ERROR)
         except Exception:
             self.log.exception("Exception in worker process")
             if not worker.booted:
-                sys.exit(self.WORKER_BOOT_ERROR)
-            sys.exit(-1)
+                os._exit(self.WORKER_BOOT_ERROR)
+            os._exit(-1)
         finally:
             self.log.info("Worker exiting (pid: %s)", worker.pid)
             try:

--- a/gunicorn/workers/base.py
+++ b/gunicorn/workers/base.py
@@ -125,7 +125,7 @@ class Worker(object):
                 os.write(self.PIPE[1], b"1")
                 self.cfg.worker_int(self)
                 time.sleep(0.1)
-                sys.exit(0)
+                os._exit(0)
 
             reloader_cls = reloader_engines[self.cfg.reload_engine]
             self.reloader = reloader_cls(extra_files=self.cfg.reload_extra_files,
@@ -195,12 +195,12 @@ class Worker(object):
         # worker_int callback
         self.cfg.worker_int(self)
         time.sleep(0.1)
-        sys.exit(0)
+        os._exit(0)
 
     def handle_abort(self, sig, frame):
         self.alive = False
         self.cfg.worker_abort(self)
-        sys.exit(1)
+        os._exit(1)
 
     def handle_error(self, req, client, addr, exc):
         request_start = datetime.now()

